### DISCUSSION TITLE
Provide an option to use `cd on quit` without requiring implementation as a shell function

### DISCUSSION THREAD

**Discussion by enoryw** [jarun/nnn#1921] [Ideas]
The feature "cd on quit" is useful for but AFAIK it's only possible to use this by running nnn in a wrapper shell function. This seems like an arbitrary requirement given that nnn is lightweight, extensible, and runs in a limited environment. How about supporting the feature natively? 

Or at least e.g. an option to output the last path on exit, so that you can do `cd "$(nnn)" in the shell. I run nnn  in an nnn wrapper script--I would rather not load the entire setup in shell config (prefer to keep it light and non-application-specific) and which could technically affect shell startup (not a real concern).

**Comment by N-R-K** [comment]
> How about supporting the feature natively?

[Not possible](https://github.com/jarun/nnn/discussions/1835#discussioncomment-8632510)

> Or at least e.g. an option to output the last path on exit

The last path is output to `$NNN_TMPFILE`, it's what the cd on quit function uses to cd on the parent shell. You can use the same logic in your custom wrapper.

> I run nnn in an nnn wrapper script

This is, I repeat, not possible. A child process (the wrapper script) cannot change the working directory of the parent process (the interactive shell).

**Reply by enoryw** [reply]
I see. The following works in the shell config (least code required):

    NNN_TMPFILE="${XDG_CONFIG_HOME:-$HOME/.config}/nnn/.lastd"

    n() {
      nnn-wrapper

      if [[ -f "$NNN_TMPFILE" ]]; then
          . "$NNN_TMPFILE"
          rm -f "$NNN_TMPFILE" > /dev/null
      fi
    }

Can you describe a little why [exporting vs. not exporting `NNN_TMPFILE`  and why it needs to be a particular path](https://github.com/jarun/nnn/blob/master/misc/quitcd/quitcd.bash_sh_zsh#L10-L14) for "cd only on ^G" to work? When I start nnn, I see its value of `NNN_TMPFILE`  is set for it whether or not `NNN_TMPFILE` was exported, so how is this different? I assume the particular path is hardcoded in the code. I understand child process cannot change the working dir of  the parent so it saves  the location for the parent to source it once the child process ends and  that `export` makes available a variable to its children processes but can't piece together how "cd only on ^G" works.

Appreciate it.

**Reply by N-R-K** [reply]
If NNN_TMPFILE is not set, nnn will write the last path only on quitting via ^G to the "hardcoded" path. If it is defined, it will write it always.

**Reply by enoryw** [reply]
Even without `export`, when I start nnn and start a shell with `!` binding, it sees this custom value, so isn't it still set in nnn? Why does nnn not just use the value of `NNN_TMPFILE` it sees and instead uses a hardcoded path?  Asking out of curiosity and to learn but also being able to define custom value of `NNN_TMPFILE` for nnn to write to (e.g. a path in tmpfs) seems to make a little more sense than to write to disk on "cd only on ^G" to hardcoded path. I prefer to define any non-essential state changes in tmpfs/memory over writing to disk. 

I only suggest using a custom value of `$NNN_TMPFILE` for "cd only on ^G" because nnn provides other variables to use like `NNN_SEL` and `NNN_FIFO`  which also carry state changes so also make sense to set to e.g. a path in tmpfs/memory.
